src/parquet/column/column-reader-test.cc

```diff
@@ -44,195 +44,135 @@ namespace test {
+      pages_.push_back(page);
+
+      num_values_ += values_per_page;
+    }
+  }
 
-  void InitReader(const ColumnDescriptor* descr) {
+  void init_reader() {
     pager_.reset(new test::MockPageReader(pages_));
-    reader_ = ColumnReader::Make(descr, std::move(pager_));
+    reader_ = ColumnReader::Make(descr_, std::move(pager_));
+  }
+
+  void check_results() {
+    vector<int32_t> vresult(num_values_, -1);
+    vector<int16_t> dresult(num_levels_, -1);
+    vector<int16_t> rresult(num_levels_, -1);
+    size_t values_read = 0;
+    size_t total_values_read = 0;
+    size_t batch_actual = 0;
+
+    Int32Reader* reader = static_cast<Int32Reader*>(reader_.get());
+    int32_t batch_size = 8;
+    size_t batch = 0;
+    // This will cover both the cases
+    // 1) batch_size < page_size (multiple ReadBatch from a single page)
+    // 2) batch_size > page_size (BatchRead from multiple pages)
+    do {
+      batch = reader->ReadBatch(batch_size, &dresult[0] + batch_actual,
+          &rresult[0] + batch_actual, &vresult[0] + total_values_read, &values_read);
+      total_values_read += values_read;
+      batch_actual += batch;
+      batch_size = batch_size * 2;
+    } while (batch > 0);
+
+    ASSERT_EQ(num_levels_, batch_actual);
+    ASSERT_EQ(num_values_, total_values_read);
+    ASSERT_TRUE(vector_equal(values_, vresult));
+    if (max_def_level_ > 0) {
+      ASSERT_TRUE(vector_equal(def_levels_, dresult));
+    }
+    if (max_rep_level_ > 0) {
+      ASSERT_TRUE(vector_equal(rep_levels_, rresult));
+    }
+  }
+
+  void execute() {
+    make_pages();
+    init_reader();
+    check_results();
   }
 
  protected:
-  std::shared_ptr<ColumnReader> reader_;
-  std::unique_ptr<PageReader> pager_;
+  int num_levels_;
+  int num_values_;
+  int num_pages_;
+  int16_t max_def_level_;
+  int16_t max_rep_level_;
+  const ColumnDescriptor *descr_;
   vector<shared_ptr<Page> > pages_;
+  std::unique_ptr<PageReader> pager_;
+  std::shared_ptr<ColumnReader> reader_;
+  vector<int32_t> values_;
+  vector<int16_t> def_levels_;
+  vector<int16_t> rep_levels_;
 };
 
-
 TEST_F(TestPrimitiveReader, TestInt32FlatRequired) {
-  vector<int32_t> values = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
-
-  std::shared_ptr<DataPage> page = MakeDataPage<Type::INT32>(values, {}, 0,
-    {}, 0);
-  pages_.push_back(page);
-
+  num_levels_ = 5000;
+  num_pages_ = 50;
+  max_def_level_ = 0;
+  max_rep_level_ = 0;
   NodePtr type = schema::Int32("a", Repetition::REQUIRED);
-  ColumnDescriptor descr(type, 0, 0);
-  InitReader(&descr);
-
-  Int32Reader* reader = static_cast<Int32Reader*>(reader_.get());
-
-  vector<int32_t> result(10, -1);
-
-  size_t values_read = 0;
-  size_t batch_actual = reader->ReadBatch(10, nullptr, nullptr,
-      &result[0], &values_read);
-  ASSERT_EQ(10, batch_actual);
-  ASSERT_EQ(10, values_read);
-
-  ASSERT_TRUE(vector_equal(result, values));
+  const ColumnDescriptor descr(type, max_def_level_, max_rep_level_);
+  descr_ = &descr;
```

Comment:
This anti-pattern suggests we should be passing `std::shared_ptr<ColumnDescriptor>` to `ColumnReader::Make`. What do you think?